src/Core/OrganizationFeatures/AuthorizationHandlers/CollectionAccessAuthorizationHandlerBase.cs

```diff
@@ -0,0 +1,99 @@
+    }
+
+    /// <summary>
+    /// Ensures the acting user is allowed to manage access permissions for the target collections.
+    /// </summary>
+    protected async Task CanManageCollectionAccessAsync(AuthorizationHandlerContext context, IAuthorizationRequirement requirement, IEnumerable<Guid> collectionIds)
+    {
+        if (!_currentContext.UserId.HasValue)
+        {
+            context.Fail();
+            return;
+        }
+
+        var distinctTargetCollectionIds = collectionIds.Distinct().ToList();
+
+        // List of collections the user is performing the operation on
+        var targetCollections = await _collectionRepository.GetManyByManyIdsAsync(distinctTargetCollectionIds);
+
+        // A target collection does not exist, fail the requirement
+        if (targetCollections.Count != distinctTargetCollectionIds.Count)
+        {
+            context.Fail();
+            return;
+        }
+
+        var userOgs = await _currentContext.OrganizationMembershipAsync(_organizationUserRepository, _currentContext.UserId.Value);
+        var distinctTargetOrganizationIds = targetCollections.Select(tc => tc.OrganizationId).Distinct().ToList();
+        var restrictedOrganizations = new List<CurrentContentOrganization>();
+
+        foreach (var orgId in distinctTargetOrganizationIds)
+        {
+            var org = userOgs.FirstOrDefault(o => orgId == o.Id);
+
+            // Acting user is not a member of the target organization, fail
+            if (org == null)
+            {
+                context.Fail();
+                return;
+            }
+
+            // Owner/Admins or users with EditAnyCollection permission can always manage collection access
+            if (
+                org.Permissions.EditAnyCollection ||
+                org.Type is OrganizationUserType.Admin or OrganizationUserType.Owner)
+            {
+                continue;
+            }
+
+            restrictedOrganizations.Add(org);
+        }
+
+        // All target collections belong to organizations the acting user is allowed to manage collection access, succeed
+        if (restrictedOrganizations.Count == 0)
+        {
+            context.Succeed(requirement);
+            return;
+        }
+
+        // List of collections the acting user is allowed to manage
+        var manageableCollections =
+            (await _collectionRepository.GetManyByUserIdAsync(_currentContext.UserId.Value))
+            .Where(c => c.Manage).ToList();
+
+        foreach (var org in restrictedOrganizations)
+        {
+            // User must have explicit "Manage" permission on the target collections for this organization
+            foreach (var targetCollection in targetCollections.Where(tc => tc.OrganizationId == org.Id))
+            {
+                // Target collection is not in the list of manageable collections, fail
+                if (!manageableCollections.Exists(c => c.Id == targetCollection.Id))
```

Comment:
There are 2 nested loops here (plus the outer loop), can it be rewritten to avoid nesting? e.g. by un-nesting the loops or by using hashsets for faster lookup

Reply:
Rewritten to remove the nested loops and instead joins/queries the lists together to find any collections that don't meet the requirements; then if any exist, fail the request.